developer-hub/src/reference_docs/safe_defensive_programming.rs

```diff
@@ -1 +1,376 @@
+//!
+//! The following code may use types from [`sp_arithmetic`].
+//!
+//! Fixed point arithmetic solves the aforementioned problems of dealing with the uncertain
+//! nature of floating point numbers. Rather than use a radix point (`0.80`), a type which _represents_
+//! a floating point number in base 10, i.e., a **fixed point number**, can be used instead.
+//!
+//! **Example - [`Perbill`](sp_arithmetic::Perbill), parts of a billion**
+//!
+#![doc = docify::embed!("./src/reference_docs/safe_defensive_programming.rs", perbill_example)]
+//!
+//! **Example - [`Percent`](sp_arithmetic::Percent), parts of a hundred**
+//!
+#![doc = docify::embed!("./src/reference_docs/safe_defensive_programming.rs", percent_example)]
+//!
+//! Note that `190 / 400 = 0.475`, and that `Percent` represents it as a _rounded down_, fixed point
+//! number (`47`). Unlike primitive types, types that implement `PerThing` will also not overflow, and
+//! are therefore safe to use. They adopt the same behavior that a saturated calculation would provide,
+//! meaning that if one is to go over "100%", it wouldn't overflow, but simply stop at the upper or
+//! lower bound.
+//!
+//! ### Using 'PerThing' In Practice
+//!
+//! [`sp_arithmetic`] contains a trait called [`PerThing`](sp_arithmetic::PerThing), allowing a custom type to be implemented specifically for fixed point arithmetic. While a number of fixed-point types are introduced, let's focus on a few specific examples that implement `PerThing`:
+//!
+//! - [`Percent`](sp_arithmetic::Percent) - parts of one hundred.
+//! - [`Permill`](sp_arithmetic::Permill) - parts of a million.
+//! - [`Perbill`](sp_arithmetic::Perbill) - parts of a billion.
+//!
+//! Because each of these implement the same trait, `PerThing`, we have access to a few widely used
+//! methods:
+//!
+//! - [`from_rational()`](sp_arithmetic::PerThing::from_rational())
+//! - [`from_percent()`](sp_arithmetic::PerThing::from_percent())
+//! - [`from_parts()`](sp_arithmetic::PerThing::from_parts())
+//!
+//! Each of these can be used to construct and represent ratios within our runtime.
+//!
+//! #### Fixed Point Arithmetic with [`PerThing`](sp_arithmetic::PerThing)
+//!
+//! As stated, one can also perform mathematics using these types directly. For example, finding the
+//! percentage of a particular item:
+//!
+#![doc = docify::embed!("./src/reference_docs/safe_defensive_programming.rs", percent_mult)]
+//!
+//! With the knowledge of how these types operate in relation to other numbers, let's explore how
+//! they're used in Substrate development.
+//!
+//! ### Fixed Point Math in Substrate Development - Further Context
+//!
+//! You will find types like [`Perbill`](sp_arithmetic::Perbill) being used often in pallet development.  [`pallet_referenda`] is a good example of a pallet 
+//! which makes good use of fixed point arithmetic to calculate 
+//!
+//! Let's examine the usage of `Perbill` and how exactly we can use it as an alternative to floating
```

Comment:
While you are at it, it would also make sense to mention the `Fixed*` types since the `PerThing` only operate in range `[0, 1]`.